### ISSUE TITLE
question: Do you use historical target variable along with historical predictor variable in you model?

### ISSUE THREAD

**Issue by xiaolongwu0713** [jdb78/pytorch-forecasting#435]
Hi,
Before I dive into this, I want to know if I can use solely historical predictor variable without the target variable to train and do the prediction?

**Comment by pnmartinez** [comment]
Hi @xiaolongwu0713 ,

I am not sure I understand your question.

- Maybe you mean predicting the target variable **using "covariates"**: other variables than the one you are trying to predict that are nevertheless related. Example: predicting temperature based on a timeseries of atmospheric pressure. 
You can do this only with [certain architectures](https://pytorch-forecasting.readthedocs.io/en/latest/models.html#id2) (mostly, everything but N-Beats, which is a single-variable model by design).

- Maybe you mean **training the model on entirely different data**; example: training on annual temperatures and use the model the predict stock market. 
You can always do that, but the performance will be generally much worse, as the model will learn the features from data that can have a very very different features than the problem you are trying to solve.

**Comment by xiaolongwu0713** [comment]
Thanks for your reply. 
I am talking about the first scenario using the covariates ( the predictor or independents variable) to predict the target (dependent variable). I think the term is nonlinear autoregressive exogenous model (NARX). But in NARX, historical target variable is also used to do the training and  prediction which is not my case. In my situation, I don't have access to the target variable, only the covariates can be used to do the training and prediction.

**Comment by pnmartinez** [comment]
Hi @xiaolongwu0713 ,

I see, that's is more specific, thank you.

In my opinion, not having access to the target variable for training is and insolvable obstacle. You cannot train for a thing you don't have data of.

If you are training with a covariate of the target variable, you will get a model that can predict the covariate. Only **if you know the specific mathematical relationship between your covariate and the unseen target variable** could you extrapolate your predicted covariate to your target variable. 

> However, **I doubt this is useful, because if you knew this mathematical relationship you wouldn't need a neural network to figure it out** in the first place.

I'll develop an example:

- Your target variable is the volume of a gas `V`,
- You train a model on a timeseries of temperatures `T`, with covariates `p`,
- You get predictions: `T_pred`, `p_pred`. 
- Through a _previously known relationship_ between variables you estimate `V_pred` from `T_pred` and `p_pred`. In this case we could use as a first approximation the ideal gas law:
```
# We use a previously known model (ideal gas law) to estimate the variable unseen by the model
pV = nRT     =>    V_pred = nR T_pred / p_pred
```

Hope that helps in some manner.

Cheers!

**Comment by doctorcolossus** [comment]
How about this? Suppose that I have access to temperature data each minute and want to predict the high and low temperature of the day. Obviously when the day is finished, and only then, are those values definitively known. Wouldn't it possible to build a model capable of making increasingly accurate predictions about those values as throughout the course of the day, based on past and recent observations? I tried something like this in pytorch_forecasting, but rather than only using the target values to calculate loss, the model seemed to train directly on those variables, as if they were known. For example, suppose that I want to validate my model with data from yesterday. In addition to DataFrame columns for timestamp and temperature, I add columns for the daily high and low using ground truth data. I want to validate my model with data from midnight to 8:00, and suppose that the actual daily high occurred later, at 13:00. The pytorch_forecasting model seems to cheat by predicting that the daily high based on the input data will just be the same as... the daily high from the input data! But I don't want to provide the daily high s input data - only as ground truth for calculating loss and backpropogating. Is there any way to do this with pytorch_forecasting?

I am new to time series prediction, but in image processing a network makes predictions based on image features to predict a categories and/or bounding boxes. Obviously the desired outputs are not fed to the network directly as inputs, but only used to evaluate and improve predictions. Isn't something analogous possible for time series data (or indeed for your example, if we didn't already know the Ideal Gas Law)? Isn't the purpose of a neural network to contruct an approximation of a potentially very complex mathematical relationship?

**Comment by doctorcolossus** [comment]
How about this? Suppose that I have access to temperature data each minute and want to predict the high and low temperature of the day. Obviously when the day is finished, and only then, are those values definitively known. Wouldn't it possible to build a model capable of making increasingly accurate predictions about those valuesthroughout the course of the day, based on past and recent observations? I tried something like this in pytorch_forecasting, but rather than only using the target values to calculate loss, the model seemed to train directly on those variables, as if they were known. For example, suppose that I want to validate my model with data from yesterday. In addition to DataFrame columns for timestamp and temperature, I add columns for the daily high and low using ground truth data. I want to validate my model with data from midnight to 8:00, and suppose that the actual daily high occurred later, at 13:00. The pytorch_forecasting model seems to cheat by predicting that the daily high based on the input data will just be the same as... the daily high from the input data! But I don't want to provide the daily high s input data - only as ground truth for calculating loss and backpropogating. Is there any way to do this with pytorch_forecasting?

I am new to time series prediction, but in image processing a network makes predictions based on image features to predict a categories and/or bounding boxes. Obviously the desired outputs are not fed to the network directly as inputs, but only used to evaluate and improve predictions. Isn't something analogous possible for time series data (or indeed for your example, if we didn't already know the Ideal Gas Law)? Isn't the purpose of a neural network to contruct an approximation of a potentially very complex mathematical relationship?

**Comment by doctorcolossus** [comment]
How about this? Suppose that I have access to temperature data each minute and want to predict the high and low temperature of the day. Obviously when the day is finished, and only then, are those values definitively known. Wouldn't it possible to build a model capable of making increasingly accurate predictions about those valuesthroughout the course of the day, based on past and recent observations? I tried something like this using pytorch_forecasting's temporal fusion transformer, but rather than only using the target values to calculate loss, the model seemed to train directly on those variables, as if they were known. For example, suppose in my DataFrame that I have columns for timestamp and temperature, as well as columns for the daily high and low. I want my model to use only the timestamp and temperature as inputs, but to use the daily high and daily low as ground truth data for calculating loss and backpropagating. The pytorch_forecasting model seems to cheat by predicting that the daily high is the same as... the daily high from the input data! Is there any way around this in pytorch_forecasting?

I am new to time series prediction, but in image processing a network makes predictions based on image features to predict a categories and/or bounding boxes. Obviously the desired outputs are not fed to the network directly as inputs, but only used to evaluate and improve predictions. Isn't something analogous possible for time series data (or indeed for your example, if we didn't already know the Ideal Gas Law)? Isn't the purpose of a neural network to contruct an approximation of a potentially very complex mathematical relationship?

**Comment by jdb78** [comment]
You can just not use the target variable as a input variable, i.e. do not include it in the `time_varying_unknown_reals`. I guess each day is a separate time series.

**Comment by doctorcolossus** [comment]
Thanks, Jan. I believe I did try that and got some error, so I had assumed it wasn't possible... It's my first time trying to use the library. I'll have another go at it and open a new issue if the error recurs and I can't seem to figure it out.

**Comment by selam08** [comment]
In such case, I guess you have to make sure your network does not use encoderNormalizer, right ?

**Comment by jannik-scheider** [comment]
> Hi @xiaolongwu0713 ,
> 
> I am not sure I understand your question.
> 
> * Maybe you mean predicting the target variable **using "covariates"**: other variables than the one you are trying to predict that are nevertheless related. Example: predicting temperature based on a timeseries of atmospheric pressure.
>   You can do this only with [certain architectures](https://pytorch-forecasting.readthedocs.io/en/latest/models.html#id2) (mostly, everything but N-Beats, which is a single-variable model by design).
> * Maybe you mean **training the model on entirely different data**; example: training on annual temperatures and use the model the predict stock market.
>   You can always do that, but the performance will be generally much worse, as the model will learn the features from data that can have a very very different features than the problem you are trying to solve.

Hi @pnmartinez, 
I found this answer from you that you posted a long time ago. I think I am currently facing the problem you described in the first scenario in my bachelor thesis. I think it is kind of a regression problem with time series data. 

I'm trying to predict Y_t given X_{t}, ..., X_{t-k}. In my timeseriesdataset I specify target=y and time_varying_unknown_reals = X.  Both X and Y are timeseries data. Is it true that during training the model learns only with the knowledge of X_{t}, ..., X_{t-k} to predict Y_t? Or does the model also include  Y_{t-1}, ..., Y_{t-k} to learn?
